In [3]:
pip install numpy pandas scikit-learn tqdm matplotlib seaborn

Defaulting to user installation because normal site-packages is not writeable

[notice] A new release of pip is available: 23.0.1 -> 24.2
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


> ### 라이브러리 호출

In [4]:
import os
from pprint import pprint

# 데이터 전처리 및 모델링
import numpy as np
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    accuracy_score,
    classification_report,
    confusion_matrix,
    f1_score,
    precision_score,
    recall_score,
)
from sklearn.model_selection import train_test_split
from tqdm import tqdm

# 시각화
import matplotlib.pyplot as plt
import seaborn as sns

#불필요한 경고 제거
import warnings
warnings.filterwarnings("ignore")

In [5]:
def read_excel_file(file_path: str, header: int = None) -> pd.DataFrame:
    csv_file = file_path.replace(".xlsx", ".csv")

    if not os.path.exists(csv_file):
        print("Converting excel to csv...")
        if header:
            df = pd.read_excel(file_path, header=header)
        else:
            df = pd.read_excel(file_path)

        df.to_csv(csv_file, index=False)
        print(f"  {file_path} -> {csv_file}")
        return df
    else:
        print(f"  Reading {csv_file}")
        return pd.read_csv(csv_file, low_memory=False)

In [6]:
# Columns 리스트 생략 없이 전부 출력하기
pd.set_option('display.max_seq_items', None)
# col 생략 없이 출력
pd.set_option('display.max_columns', None)

In [13]:
RANDOM_STATE = 110
dispenser1_df_train = pd.read_csv(os.path.join("/home/elicer/LG/dispenser1_df_train.csv"), index_col=0)
dispenser2_df_train = pd.read_csv(os.path.join("/home/elicer/LG/dispenser2_df_train.csv"), index_col=0)
dispenser1_df_test = pd.read_csv(os.path.join("/home/elicer/LG/dispenser1_df_test.csv"), index_col=0)
dispenser2_df_test = pd.read_csv(("/home/elicer/LG/dispenser2_df_test.csv"), index_col=0)

In [18]:
dispenser1_df_test

CURE SPEED Collect Result_Dam  \
2                                 70   
4                                 70   
5                                 70   
6                                 70   
7                                 70   
...                              ...   
17353                             70   
17355                             70   
17358                             70   
17359                             70   
17360                             70   

       DISCHARGED SPEED OF RESIN Collect Result_Dam  \
2                                                10   
4                                                16   
5                                                10   
6                                                10   
7                                                10   
...                                             ...   
17353                                            10   
17355                                            10   
17358                                            16   
17359                                            10   
17360                                            10   

       DISCHARGED TIME OF RESIN(Stage1) Collect Result_Dam  \
2                                                    9.7     
4                                                   13.2     
5                                                    9.6     
6                                                    9.6     
7                                                    9.7     
...                                                  ...     
17353                                                9.6     
17355                                               21.3     
17358                                               13.2     
17359                                                9.7     
17360                                               17.0     

       DISCHARGED TIME OF RESIN(Stage2) Collect Result_Dam  \
2                                                    4.9     
4                                                    7.5     
5                                                    4.9     
6                                                    4.9     
7                                                    4.9     
...                                                  ...     
17353                                                3.9     
17355                                                5.5     
17358                                                6.6     
17359                                                3.9     
17360                                                4.9     

       DISCHARGED TIME OF RESIN(Stage3) Collect Result_Dam  \
2                                                    9.7     
4                                                   13.2     
5                                                    9.7     
6                                                    9.7     
7                                                    9.7     
...                                                  ...     
17353                                                9.7     
17355                                               21.3     
17358                                               13.2     
17359                                                9.7     
17360                                               17.1     

       Dispense Volume(Stage1) Collect Result_Dam  \
2                                            0.67   
4                                            0.92   
5                                            0.67   
6                                            0.67   
7                                            0.67   
...                                           ...   
17353                                        0.67   
17355                                        1.49   
17358                                        1.45   
17359                                        0.67   
17360                                        1.

In [ ]:
pip install optuna

Defaulting to user installation because normal site-packages is not writeable

[notice] A new release of pip is available: 23.0.1 -> 24.2
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [ ]:
pip install lightgbm --install-option=--gpu

Defaulting to user installation because normal site-packages is not writeable
DEPRECATION: --install-option is deprecated because it forces pip to use the 'setup.py install' command which is itself deprecated. pip 23.1 will enforce this behaviour change. A possible replacement is to use --config-settings. Discussion can be found at https://github.com/pypa/pip/issues/11358
DEPRECATION: --no-binary currently disables reading from the cache of locally built wheels. In the future --no-binary will not influence the wheel cache. pip 23.1 will enforce this behaviour change. A possible replacement is to use the --no-cache-dir option. You can use the flag --use-feature=no-binary-enable-wheel-cache to test the upcoming behaviour. Discussion can be found at https://github.com/pypa/pip/issues/11453

[notice] A new release of pip is available: 23.0.1 -> 24.2
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [ ]:
pip install catboost

Defaulting to user installation because normal site-packages is not writeable

[notice] A new release of pip is available: 23.0.1 -> 24.2
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [ ]:
import pandas as pd
from catboost import CatBoostClassifier
from sklearn.model_selection import train_test_split
import optuna

In [ ]:
import optuna
from catboost import CatBoostClassifier
from sklearn.metrics import f1_score
from sklearn.model_selection import cross_val_score, StratifiedKFold
import numpy as np

# Feature와 Target 분리
X = dispenser1_df_train.drop(columns='target')
y = dispenser1_df_train['target']

# Optuna Objective Function 정의
def objective(trial):
    param = {
        'learning_rate': trial.suggest_loguniform('learning_rate', 1e-2, 0.1),
        'depth': trial.suggest_int('depth', 7, 10),
        'l2_leaf_reg': trial.suggest_loguniform('l2_leaf_reg', 1e-3, 0.1),
        'n_estimators': trial.suggest_int('n_estimators', 300, 500),
        'task_type': 'GPU'
    }

    model = CatBoostClassifier(
        **param,
        loss_function='Logloss',
        eval_metric='F1',
        cat_features=["Chamber Temp. Judge Value_AutoClave", "Change in the Column"],
        verbose=0
    )

    # Stratified K-Fold Cross Validation
    skf = StratifiedKFold(n_splits=5, shuffle=True, random_state= 110)
    f1_scores = []

    for train_index, valid_index in skf.split(X, y):
        X_train, X_valid = X.iloc[train_index], X.iloc[valid_index]
        y_train, y_valid = y.iloc[train_index], y.iloc[valid_index]
        
        model.fit(X_train, y_train, eval_set=(X_valid, y_valid), early_stopping_rounds=200, verbose=False)
        
        # Validation prediction and F1 score calculation
        y_valid_pred = model.predict(X_valid)
        f1 = f1_score(y_valid, y_valid_pred)
        f1_scores.append(f1)
    
    # Return the mean F1 score across all folds
    return np.mean(f1_scores)

# Optuna Study 생성 및 최적화
study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials= 200)

# 최적의 하이퍼파라미터 출력
print("Best Parameters:", study.best_params)
print("Best F1 Score:", study.best_value)


In [ ]:
# 최적의 하이퍼파라미터로 모델 학습
best_params = study.best_params
best_model = CatBoostClassifier(
    **best_params,
    loss_function='Logloss',
    eval_metric='F1',
    cat_features=["Chamber Temp. Judge Value_AutoClave", "Change in the Column"],
    verbose=0
)

# 전체 훈련 데이터에서 모델을 재학습
best_model.fit(X, y, early_stopping_rounds=200, verbose=False)

# 교차 검증으로 최종 F1 Score 계산
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
final_f1_scores = []

for train_index, valid_index in skf.split(X, y):
    X_train, X_valid = X.iloc[train_index], X.iloc[valid_index]
    y_train, y_valid = y.iloc[train_index], y.iloc[valid_index]
    
    best_model.fit(X_train, y_train, eval_set=(X_valid, y_valid), early_stopping_rounds=150, verbose=False)
    
    # Validation prediction and F1 score calculation
    y_valid_pred = best_model.predict(X_valid)
    f1 = f1_score(y_valid, y_valid_pred)
    final_f1_scores.append(f1)

# 최종 F1 Score 출력
final_f1_score = np.mean(final_f1_scores)
print("Best Parameters:", best_params)
print("Final F1 Score:", final_f1_score)

Best Parameters: {'learning_rate': 0.08288130501424704, 'depth': 7, 'l2_leaf_reg': 0.006569838906834511, 'n_estimators': 393}
Final F1 Score: 0.1316947302121257


## XGBoost

In [ ]:
pip install xgboost

Defaulting to user installation because normal site-packages is not writeable
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 153.9/153.9 MB 11.6 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 190.9/190.9 MB 7.7 MB/s eta 0:00:0000:0100:01

[notice] A new release of pip is available: 23.0.1 -> 24.2
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [ ]:
pip install --upgrade xgboost

Defaulting to user installation because normal site-packages is not writeable

[notice] A new release of pip is available: 23.0.1 -> 24.2
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [ ]:
pip install --upgrade scikit-learn

Defaulting to user installation because normal site-packages is not writeable

[notice] A new release of pip is available: 23.0.1 -> 24.2
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [ ]:
# 'Chamber Temp. Judge Value_AutoClave' 칼럼을 'category' 타입으로 변환
dispenser2_df_train["Chamber Temp. Judge Value_AutoClave"] = dispenser2_df_train["Chamber Temp. Judge Value_AutoClave"].astype('category')

In [ ]:
dispenser2_df_train["Change in the Column"] = dispenser2_df_train["Change in the Column"].map({0: 'NP', 1: 'PE'})

# 피쳐 타입을 category로 변경
dispenser2_df_train["Change in the Column"] = dispenser2_df_train["Change in the Column"].astype('category')

In [ ]:
# 카테고리형 칼럼에 대해 원핫 인코딩 수행
dispenser2_df_train = pd.get_dummies(dispenser2_df_train, drop_first=True)

In [ ]:
import optuna
from xgboost import XGBClassifier
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import StratifiedKFold
import numpy as np

# Feature와 Target 분리
X = dispenser2_df_train.drop(columns='target')
y = dispenser2_df_train['target']

# Optuna Objective Function 정의
def objective(trial):
    param = {
        'learning_rate': trial.suggest_loguniform('learning_rate', 1e-2, 0.1),
        'max_depth': trial.suggest_int('max_depth', 3, 10),
        'min_child_weight': trial.suggest_int('min_child_weight', 1, 10),
        'gamma': trial.suggest_loguniform('gamma', 1e-3, 0.1),
        'subsample': trial.suggest_uniform('subsample', 0.5, 1.0),
        'colsample_bytree': trial.suggest_uniform('colsample_bytree', 0.5, 1.0),
        'alpha': trial.suggest_loguniform('alpha', 1e-3, 0.1),
        'lambda': trial.suggest_loguniform('lambda', 1e-3, 0.1),
        'scale_pos_weight': trial.suggest_loguniform('scale_pos_weight', 1, 10),
        'n_estimators': trial.suggest_int('n_estimators', 300, 500),
        'objective': 'binary:logistic',
        'eval_metric': 'auc',  # AUC로 변경
        'tree_method': 'gpu_hist',  # GPU 지원
    }

    model = XGBClassifier(
        **param,
        use_label_encoder=False,
        verbose=0
    )

    # Stratified K-Fold Cross Validation
    skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=110)
    auc_scores = []

    for train_index, valid_index in skf.split(X, y):
        X_train, X_valid = X.iloc[train_index], X.iloc[valid_index]
        y_train, y_valid = y.iloc[train_index], y.iloc[valid_index]
        
        model.fit(
            X_train, y_train, 
            eval_set=[(X_valid, y_valid)],
            verbose=False
        )
        
        # Validation prediction and AUC score calculation
        y_valid_pred_prob = model.predict_proba(X_valid)[:, 1]
        auc = roc_auc_score(y_valid, y_valid_pred_prob)
        auc_scores.append(auc)
    
    # Return the mean AUC score across all folds
    return np.mean(auc_scores)

# Optuna Study 생성 및 최적화
study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials=300)

# 최적의 하이퍼파라미터 출력
print("Best Parameters:", study.best_params)
print("Best AUC Score:", study.best_value)

[I 2024-08-23 04:46:10,237] A new study created in memory with name: no-name-21a73c83-90ae-4f5a-a4a7-859a5a02f6fc
[I 2024-08-23 04:46:15,863] Trial 0 finished with value: 0.6327642079995994 and parameters: {'learning_rate': 0.07719187185934523, 'max_depth': 3, 'min_child_weight': 2, 'gamma': 0.01097621205455851, 'subsample': 0.5035597400677524, 'colsample_bytree': 0.7564241399492424, 'alpha': 0.004530381807753004, 'lambda': 0.0035835399934785236, 'scale_pos_weight': 5.281070051583352, 'n_estimators': 447}. Best is trial 0 with value: 0.6327642079995994.
[I 2024-08-23 04:46:22,957] Trial 1 finished with value: 0.6363295204362485 and parameters: {'learning_rate': 0.057792486473125536, 'max_depth': 10, 'min_child_weight': 8, 'gamma': 0.00945734758551075, 'subsample': 0.5786005225426665, 'colsample_bytree': 0.6209222465416628, 'alpha': 0.007147246343258395, 'lambda': 0.0015579971760909979, 'scale_pos_weight': 5.913646135806862, 'n_estimators': 316}. Best is trial 1 with value: 0.6363295204

Best Parameters: {'learning_rate': 0.01045847023114772, 'max_depth': 5, 'min_child_weight': 6, 'gamma': 0.024093374485715007, 'subsample': 0.7945419264433988, 'colsample_bytree': 0.5181957466328934, 'alpha': 0.044481412158161995, 'lambda': 0.003927752204104039, 'scale_pos_weight': 2.5424860472947346, 'n_estimators': 415}
Best AUC Score: 0.6576491337331722


In [ ]:
# 최적의 하이퍼파라미터로 모델 학습
best_params = study.best_params
best_model = XGBClassifier(
    **best_params,
    use_label_encoder=False,
    verbose=0
)

# 전체 훈련 데이터에서 모델을 재학습
best_model.fit(X, y, eval_set=[(X, y)], verbose=False)

# 교차 검증으로 최종 F1 Score 계산
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
final_f1_scores = []

for train_index, valid_index in skf.split(X, y):
    X_train, X_valid = X.iloc[train_index], X.iloc[valid_index]
    y_train, y_valid = y.iloc[train_index], y.iloc[valid_index]
    
    best_model.fit(X_train, y_train, eval_set=[(X_valid, y_valid)], verbose=False)
    
    # Validation prediction and F1 score calculation
    y_valid_pred = best_model.predict(X_valid)
    f1 = f1_score(y_valid, y_valid_pred)
    final_f1_scores.append(f1)

# 최종 F1 Score 출력
final_f1_score = np.mean(final_f1_scores)
print("Best Parameters:", best_params)
print("Final F1 Score:", final_f1_score)

Best Parameters: {'learning_rate': 0.01045847023114772, 'max_depth': 5, 'min_child_weight': 6, 'gamma': 0.024093374485715007, 'subsample': 0.7945419264433988, 'colsample_bytree': 0.5181957466328934, 'alpha': 0.044481412158161995, 'lambda': 0.003927752204104039, 'scale_pos_weight': 2.5424860472947346, 'n_estimators': 415}
Final F1 Score: 0.1752927884921643


> Optuna 결과

<파생변수 칼럼 변화 있는 거 1개만 했을 때>
- dispenser 1 : catboost => {'learning_rate': 0.08995613012147986, 'depth': 7, 'l2_leaf_reg': 0.007655807591808423, 'n_estimators': 454} ==> Final F1 Score: 0.128
- dispenser 2 : catboost => {'learning_rate': 0.09991283555793407, 'depth': 9, 'l2_leaf_reg': 0.006938972685745867, 'n_estimators': 398} ==> Final F1 Score: 0.19227

<파생변수 categorical로 넣어주었을때>
- dispenser1 {'learning_rate': 0.08288130501424704, 'depth': 7, 'l2_leaf_reg': 0.006569838906834511, 'n_estimators': 393} ==> Final F1 Score: 0.13169
==> 이렇게 가자.

<파생변수 많이 만들었을 때>


<XGBOOST>
- dispenser 2 : xgboost => {'learning_rate': 0.01045847023114772, 'max_depth': 5, 'min_child_weight': 6, 'gamma': 0.024093374485715007, 'subsample': 0.7945419264433988, 'colsample_bytree': 0.5181957466328934, 'alpha': 0.044481412158161995, 'lambda': 0.003927752204104039, 'scale_pos_weight': 2.5424860472947346, 'n_estimators': 415} ==> Final F1 Score: 0.175292


>>>>>>>> CatBoost => 최종 분류기
>>>>>>>>언더 샘플링

1. 분류기의 예측 확률로 없애기
- 임계값 통일하고, 분류기 성능 보고 이후에는 가장 높은 확률로 normal이라고 예측한 값들 => 2500개 정도를 지울 예정
-무조건 AbNormal인 데이터는 확률을 1이라고 두고 나머지는 다 모델의 예측확률로 새로운 피쳐로 넣어준다. 이러고 어떤 임계값에 따라 0에 엄청 가까운 데이터들을 제거해준다.

2. Tomek links(경계선에서 Normal인 것들을 지워버리는 기술=> 경계선 근처라 효과가 미미)

3. CNN(Condensed Nearest Neighbor) => 경계면 안 쪽의 것들을 지우는 기술

4. OSS(One-Sided Selection) => 경계면의 소수클래스를 제거 ==> 이건 조금 아닌 거 같은디..?
OSS 방법은 Tomek link + CNN이라고 생각하면 됩니다. Tomek link로 다수 범주의 데이터를 제거 후 CNN을 사용하여 다시 한 번 언더 샘플링 수행합니다. Tomek link로 클래스의 경계면에 있는 데이터를 제거하고 CNN으로는 경계면에서 멀리 떨어진 다수 범주의 데이터를 제거하게 됩니다. 이 기법은 Tomek link와 CNN의 한계를 상호 보완했다는 장점이 있습니다. 

- 2. scaling 해서 auto encoder로 autoclave는 남겨두고 dam, fill1, fill2 차원 축소 시키고
- 3. smote로 데이터 늘려서 예측 수행
- 2,3의 순서는 고민을 해봐야 할듯

In [ ]:
from sklearn.metrics import f1_score

# 학습된 최적의 모델로 validation set 예측 (확률 예측)
y_valid_pred_proba = best_model.predict_proba(X_valid)[:, 1]  # 긍정 클래스(1)의 확률

# 임계값을 조정하여 예측 생성
threshold = 0.06
y_valid_pred = (y_valid_pred_proba >= threshold).astype(int)

# F1 Score로 성능 평가
f1 = f1_score(y_valid, y_valid_pred)
print(f"F1 Score (threshold={threshold}):", f1)

In [ ]:
# 테스트 데이터 예측
X_test = dispenser1_df_test  # 테스트 데이터 준비
y_test_pred = best_model.predict(X_test)

# 결과 저장 (예를 들어 CSV 파일로 저장)
pd.DataFrame(y_test_pred, columns=['Prediction']).to_csv("test_predictions.csv", index=False)

print("Test predictions saved to 'test_predictions.csv'")

## Tomek link => under sampling

In [122]:
 train_x, train_y, val_x, val_y= model_train_test_split(dispenser1_df_train)

In [124]:
pip install imbalanced-learn

Defaulting to user installation because normal site-packages is not writeable

[notice] A new release of pip is available: 23.0.1 -> 24.2
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [128]:
# Tomek Links는 categorical 변수 처리 불가
# Chamber Temp. Judge Value_AutoClave 칼럼 값을 변경
train_x['Chamber Temp. Judge Value_AutoClave'] = train_x['Chamber Temp. Judge Value_AutoClave'].map({'OK': 0, 'NG': 1})

# 데이터 타입을 정수형으로 변환
train_x['Chamber Temp. Judge Value_AutoClave'] = train_x['Chamber Temp. Judge Value_AutoClave'].astype(int)


In [134]:
from imblearn.under_sampling import TomekLinks

# Tomek Links 객체 생성
tomek_links = TomekLinks(sampling_strategy='majority')

# Tomek Links를 이용한 언더샘플링 수행
train_x_resampled, train_y_resampled = tomek_links.fit_resample(train_x, train_y)

# 결과 확인
print("Original train_x shape:", train_x.shape)
print("Resampled train_x shape:", train_x_resampled.shape)
print("Original target class distribution:", train_y.value_counts())
print("Resampled target class distribution:", pd.Series(train_y_resampled).value_counts())


Original train_x shape: (17521, 80)
Resampled train_x shape: (17009, 80)
Original target class distribution: target
0    16481
1     1040
Name: count, dtype: int64
Resampled target class distribution: target
0    15969
1     1040
Name: count, dtype: int64


In [135]:
from imblearn.under_sampling import TomekLinks

# Tomek Links 객체 생성
tomek_links = TomekLinks(sampling_strategy='majority')

# Tomek Links를 이용한 언더샘플링 수행
train_x_resampled2, train_y_resampled2 = tomek_links.fit_resample(train_x_resampled, train_y_resampled)

# 결과 확인
print("Original train_x shape:", train_x_resampled.shape)
print("Resampled train_x shape:", train_x_resampled2.shape)
print("Original target class distribution:", train_y_resampled.value_counts())
print("Resampled target class distribution:", pd.Series(train_y_resampled2).value_counts())

Original train_x shape: (17009, 80)
Resampled train_x shape: (16805, 80)
Original target class distribution: target
0    15969
1     1040
Name: count, dtype: int64
Resampled target class distribution: target
0    15765
1     1040
Name: count, dtype: int64


In [136]:
from imblearn.under_sampling import OneSidedSelection

# OSS 객체 생성
oss = OneSidedSelection(sampling_strategy='majority', random_state=42)

# OSS를 이용한 언더샘플링 수행
train_x_resampled_OSS, train_y_resampled_OSS = oss.fit_resample(train_x, train_y)

# 결과 확인
print("Original train_x shape:", train_x.shape)
print("Resampled train_x shape:", train_x_resampled_OSS.shape)
print("Original target class distribution:", train_y.value_counts())
print("Resampled target class distribution:", pd.Series(train_y_resampled_OSS).value_counts())


Original train_x shape: (17521, 80)
Resampled train_x shape: (16981, 80)
Original target class distribution: target
0    16481
1     1040
Name: count, dtype: int64
Resampled target class distribution: target
0    15941
1     1040
Name: count, dtype: int64


> 모델 훈련 시작

In [ ]:
# dispenser 1
cat_model_dispenser1 = cat_model_fit(train_x_dispenser1, train_y_dispenser1, val_x_dispenser1, val_y_dispenser1)

In [ ]:
# dispenser 2
cat_model_dispenser2 = cat_model_fit(train_x_dispenser2, train_y_dispenser2, val_x_dispenser2, val_y_dispenser2)

> **여기서부터는 prediction threshold, model parameter 조정해보는거야**

In [ ]:
# 예측 확률에 AbNormal로 뺴두었던 index 추가하기

def add_unconditionally_AbNormal_dispenser1(pred_proba_numpy_array, val_x, submission_abnormal_row_index):
  # y_pred_proba1을 DataFrame으로 변환
  y_pred_proba_df = pd.DataFrame(y_pred_proba1, index=val_x_dispenser1.index, columns=['probability'])

  # Submission_AbNormal_row_index의 인덱스를 y_pred_proba_df에 추가하고, 값을 1로 설정
  for idx in submission_abnormal_row_index:
    if idx not in y_pred_proba_df.index:
      y_pred_proba_df.loc[idx] = 1.0  # 인덱스가 없으면 추가하고 값은 1.0으로 설정

  # y_pred_proba_df을 다시 Numpy로 변경
  modified_pred_proba_array = y_pred_proba_df.sort_index()['probability'].values

  return modified_pred_proba_array

In [ ]:
# validatoin_y에 AbNormal 추가하기
def add_abnormal_rows(val_y, submission_abnormal_row_index):
  # val_y의 기존 인덱스
  existing_indices = val_y.index.tolist()
  # Submission_AbNormal_row_index에서 val_y의 인덱스와 겹치지 않는 인덱스를 필터링
  new_indices = [idx for idx in submission_abnormal_row_index if idx not in existing_indices]

  # 기존 데이터프레임에 새로운 인덱스를 추가하고 'target' 값을 1로 설정
  for idx in new_indices:
    val_y.loc[idx] = 1
  # 인덱스를 정렬
  updated_val_y = val_y.sort_index()

  # 인덱스를 정렬
  updated_val_y = updated_val_y.values

  return updated_val_y

In [ ]:
# 예측 - dispenser1
y_pred_proba1 = cat_model_dispenser1.predict_proba(val_x_dispenser1)[:, 1]
y_pred_proba1 = add_unconditionally_AbNormal_dispenser1(y_pred_proba1, val_x_dispenser1, Submission_AbNormal_row_index)
y_pred1 = (y_pred_proba1 > 0.3).astype(int)
updated_val_y = add_abnormal_rows(val_y_dispenser1, Submission_AbNormal_row_index)
# F1 스코어 계산
f1 = f1_score(updated_val_y, y_pred1)
print(f'F1 Score: {f1:.4f}')

In [ ]:
# 예측 확률에 AbNormal로 뺴두었던 index 추가하기

def add_unconditionally_AbNormal_dispenser2(pred_proba_numpy_array, val_x, submission_abnormal_row_index):
  # y_pred_proba1을 DataFrame으로 변환
  y_pred_proba_df = pd.DataFrame(y_pred_proba2, index=val_x_dispenser2.index, columns=['probability'])

  # Submission_AbNormal_row_index의 인덱스를 y_pred_proba_df에 추가하고, 값을 1로 설정
  for idx in submission_abnormal_row_index:
    if idx not in y_pred_proba_df.index:
      y_pred_proba_df.loc[idx] = 1.0  # 인덱스가 없으면 추가하고 값은 1.0으로 설정

  # y_pred_proba_df을 다시 Numpy로 변경
  modified_pred_proba_array = y_pred_proba_df.sort_index()['probability'].values

  return modified_pred_proba_array

In [ ]:
# 예측 - dispenser2
y_pred_proba2 = cat_model_dispenser2.predict_proba(val_x_dispenser2)[:, 1]
y_pred_proba2 = add_unconditionally_AbNormal_dispenser2(y_pred_proba2, val_x_dispenser2, Submission_AbNormal_row_index)
y_pred2 = (y_pred_proba2 > 0.3).astype(int)
updated_val_y_2 = add_abnormal_rows(val_y_dispenser2, Submission_AbNormal_row_index)
# F1 스코어 계산
f2 = f1_score(updated_val_y_2, y_pred2)
print(f'F1 Score: {f2:.4f}')

### Analizing Prediction Result_TRAIN

In [ ]:
train_x_dispenser1 = convert_to_category(train_x_dispenser1)

# 특성 중요도 계산 및 출력
category_columns = train_x_dispenser1.select_dtypes(include='category').columns.tolist()
feature_importances = cat_model_dispenser1.get_feature_importance(Pool(train_x_dispenser1, label=train_y_dispenser1, cat_features=category_columns))

# 특성 중요도를 내림차순으로 정렬하고 상위 30개의 인덱스를 선택
sorted_idx = np.argsort(feature_importances)[::-1]  # 내림차순으로 정렬된 인덱스
top_40_idx = sorted_idx[:40]  # 상위 30개의 인덱스 선택

# 상위 30개의 특성과 중요도만 선택
top_40_features = train_x_dispenser1.columns[top_40_idx]
top_40_importances = feature_importances[top_40_idx]

# 상위 30개 특성에 대해 시각화
plt.figure(figsize=(10, 10))
plt.barh(top_40_features, top_40_importances)
plt.xlabel('Feature Importance')
plt.ylabel('Features')
plt.title('Top 40 Feature Importance in CatBoost Model')
plt.show()

### PREDICT & SCORE => Test data

In [ ]:
dispenser1_df_test.shape[0]

In [ ]:
dispenser2_df_test.shape[0]

In [ ]:
y_pred_proba1_test = cat_model_dispenser1.predict_proba(dispenser1_df_test)[:, 1]
y_pred1_test = (y_pred_proba1_test > 0.11).astype(int)

y_pred_proba2_test = cat_model_dispenser2.predict_proba(dispenser2_df_test)[:, 1]
y_pred2_test = (y_pred_proba2_test > 0.11).astype(int)

# submission에 대입
submission_df.loc[dispenser1_df_test.index, 'target'] = y_pred1_test
submission_df.loc[dispenser2_df_test.index, 'target'] = y_pred2_test
submission_df['target'] = submission_df['target'].map({0: 'Normal', 1: 'AbNormal'})
submission_df.loc[Submission_AbNormal_row_index_test, 'target'] = "AbNormal"

In [ ]:
# predict 임계값 : 0.11, depth : 5 , earlystopping : 40  ==> 0.185
submission_df["target"].value_counts()
# 15:1

In [ ]:
submission_df.to_csv("submission.csv", index=False)

In [ ]:
train_data["target"].value_counts()
# 16 : 1

> #### 데이터를 고쳤는데도 점수가 거의 오르지 않은 것으로 보아, 확실히 AbNormal인 데이터들이 적은데 빼니까 더 감지가 어려운듯함. 처음부터 제외하고 학습을 시키기 보다는 같이 학습을 시키는 것이 좋아보임

### 1. undersampling